# Importing the needed packages

In [1]:
import pandas as pd
pd.options.mode.chained_assignment = None  # default='warn'

# Data Access and Profiling:
###     Importing the Restaurant Menu and Nutrition dataset

In [2]:
df= pd.read_csv ('/Users/buzz/Downloads/Business Info Project/Menu.csv') 
df

/var/folders/sk/ykmmm_7n6dj04ryb4qcvnbpc0000gn/T/ipykernel_34742/134697359.py:1: DtypeWarning: Columns (9,40) have mixed types. Specify dtype option on import or set low_memory=False.
  df= pd.read_csv ('/Users/buzz/Downloads/Business Info Project/Menu.csv')


Menu_Item_ID  Year                               Restaurant_Item_Name  \
0             35005  2017                         7 Eleven Mocha Iced Coffee   
1             35008  2017                7 Eleven French Vanilla Iced Coffee   
2             35027  2017                 7 Eleven French Vanilla Cappuccino   
3             35028  2017                          7 Eleven Peppermint Mocha   
4             35029  2017                       7 Eleven Pumpkin Spice Latte   
...             ...   ...                                                ...   
65214        118344  2018               Einstein Bros Frozen Mocha, 24 fl oz   
65215        153014  2018  Dunkin' Donuts Sweet Black Pepper Bacon Wake U...   
65216        174902  2018                  LongHorn Steakhouse Whiskey Smash   
65217        131673  2018                    Golden Corral Chicken Enchilada   
65218        136001  2018                 Del Taco Street Tacos, Carne Asada   

                restaurant  Restaurant_ID  \
0                 7 Eleven              1   
1                 7 Eleven              1   
2                 7 Eleven              1   
3                 7 Eleven              1   
4                 7 Eleven              1   
...                    ...            ...   
65214        Einstein Bros             31   
65215       Dunkin' Donuts             30   
65216  LongHorn Steakhouse             53   
65217        Golden Corral             37   
65218             Del Taco             27   

                                   Item_Name  \
0                          Mocha Iced Coffee   
1                 French Vanilla Iced Coffee   
2                  French Vanilla Cappuccino   
3                           Peppermint Mocha   
4                        Pumpkin Spice Latte   
...                                      ...   
65214                 Frozen Mocha, 24 fl oz   
65215  Sweet Black Pepper Bacon Wake Up Wrap   
65216                          Whiskey Smash   
65217                      Chicken Enchilada   
65218              Street Tacos, Carne Asada   

                                        Item_Description Food_Category  \
0        Mocha Iced Coffee, Chillers Iced Coffee, Drinks     Beverages   
1      French Vanilla Iced Coffee, Chillers Iced Coff...     Beverages   
2      French Vanilla Cappuccino, Coffee, Drinks, Fla...     Beverages   
3                      Peppermint Mocha, Coffee, 8 fl oz     Beverages   
4                   Pumpkin Spice Latte, Coffee, 8 fl oz     Beverages   
...                                                  ...           ...   
65214  Frozen Mocha w/ 2% Milk, 24 fl oz, Classic, Be...     Beverages   
65215  Sweet Black Pepper Bacon Wake Up Wrap, Breakfa...    Sandwiches   
65216  Whiskey Smash w/ Jim Beam Bourbon, Fresh Squee...     Beverages   
65217                Chicken Enchilada, Chicken & Turkey       Entrees   
65218  Street Tacos in 2 Tortillas w/ Carne Asada Ste...    Sandwiches   

       Serving_Size Serving_Size_text  ... Sodium_text Potassium_text  \
0               NaN               NaN  ...         NaN            NaN   
1               NaN               NaN  ...         NaN            NaN   
2               8.0               NaN  ...         NaN            NaN   
3               8.0               NaN  ...         NaN            NaN   
4               8.0               NaN  ...         NaN            NaN   
...             ...               ...  ...         ...            ...   
65214          24.0               NaN  ...         NaN            NaN   
65215           NaN               NaN  ...         NaN            NaN   
65216           NaN               NaN  ...         NaN            NaN   
65217         127.0               NaN  ...         NaN            NaN   
65218         113.0               NaN  ...         NaN            NaN   

       Carbohydrates_text  Protein_text  Sugar_text  Dietary_Fiber_text  \
0                     NaN           NaN         NaN                 NaN   
1           

###  * The number of observations and the number of variables in the dataset.

In [4]:
# Checking shape of dataframe
df.shape

(65219, 49)

###  *  The name and data type of each data column.


In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65219 entries, 0 to 65218
Data columns (total 49 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Menu_Item_ID            65219 non-null  int64  
 1   Year                    65219 non-null  int64  
 2   Restaurant_Item_Name    65219 non-null  object 
 3   restaurant              65219 non-null  object 
 4   Restaurant_ID           65219 non-null  int64  
 5   Item_Name               65219 non-null  object 
 6   Item_Description        65219 non-null  object 
 7   Food_Category           65219 non-null  object 
 8   Serving_Size            26899 non-null  float64
 9   Serving_Size_text       39 non-null     object 
 10  Serving_Size_Unit       26927 non-null  object 
 11  Serving_Size_household  15238 non-null  object 
 12  Calories                55315 non-null  float64
 13  Total_Fat               54846 non-null  float64
 14  Saturated_Fat           54143 non-null

### Checking unique values and count in each column

In [7]:
cols = df.columns
cols

Index(['Menu_Item_ID', 'Year', 'Restaurant_Item_Name', 'restaurant',
       'Restaurant_ID', 'Item_Name', 'Item_Description', 'Food_Category',
       'Serving_Size', 'Serving_Size_text', 'Serving_Size_Unit',
       'Serving_Size_household', 'Calories', 'Total_Fat', 'Saturated_Fat',
       'Trans_Fat', 'Cholesterol', 'Sodium', 'Potassium', 'Carbohydrates',
       'Protein', 'Sugar', 'Dietary_Fiber', 'Calories_100g', 'Total_Fat_100g',
       'Saturated_Fat_100g', 'Trans_Fat_100g', 'Cholesterol_100g',
       'Sodium_100g', 'Potassium_100g', 'Carbohydrates_100g', 'Protein_100g',
       'Sugar_100g', 'Dietary_Fiber_100g', 'Calories_text', 'Total_Fat_text',
       'Saturated_Fat_text', 'Trans_Fat_text', 'Cholesterol_text',
       'Sodium_text', 'Potassium_text', 'Carbohydrates_text', 'Protein_text',
       'Sugar_text', 'Dietary_Fiber_text', 'Kids_Meal', 'Limited_Time_Offer',
       'Regional', 'Shareable'],
      dtype='object')

In [9]:

# for each column
for col in cols:
    print(col)

    # get a list of unique values
    unique = df[col].unique()
    print(unique, '\n*************************************\n\n')


Menu_Item_ID
[ 35005  35008  35027 ... 172124 177768 174902] 
*************************************


Year
[2017 2018] 
*************************************


Restaurant_Item_Name
['7 Eleven Mocha Iced Coffee' '7 Eleven French Vanilla Iced Coffee'
 '7 Eleven French Vanilla Cappuccino' ...
 'Wawa Butter, Regular, for Junior Hot Hoagies'
 "Dunkin' Donuts Sweet Black Pepper Bacon Wake Up Wrap"
 'LongHorn Steakhouse Whiskey Smash'] 
*************************************


restaurant
['7 Eleven' "Applebee's" 'Chuck E. Cheese' "Arby's" 'Red Robin'
 "BJ's Restaurant & Brewhouse" 'Bonefish Grill' "Auntie Anne's"
 "Captain D's" 'Wawa' 'Sbarro' 'Bojangles' "O'Charley's" 'Baskin Robbins'
 'Bob Evans' 'Dominos' "Jimmy John's" "Friendly's" 'Boston Market'
 'Burger King' 'California Pizza Kitchen' "Carl's Jr."
 "Carrabba's Italian Grill" "Casey's General Store"
 "Checker's Drive-In/Rallys" 'Chick-Fil-A' "Chili's" 'Chipotle'
 "Church's Chicken" 'Olive Garden' "Hardee's" "Ci Ci's Pizza" "Culver's"
 '

 2.080e+02 3.730e+02 1.610e+02 2.040e+02 2.110e+02] 
*************************************


Sodium_100g
[  nan   84.   79. ... 1012.  957.  825.] 
*************************************


Potassium_100g
[ nan   1.   0.   2. 162. 166. 154. 171. 158. 198. 163. 186. 210. 160.
 183. 100. 229. 192. 138. 144. 262. 148. 155. 151. 145. 140. 146. 276.
 281. 412. 411. 408. 320. 199. 195. 203. 128. 250. 249. 110.  93. 280.
  77.  42. 429. 161.  71.  60. 214.  66. 278. 136. 147.  83. 111. 208.
  92. 149. 222. 168. 152. 401. 404. 417. 109.  88. 125. 167.  91.  36.
 180.  11.  17.  18.  16.  15.  61.  48.  53.  70. 194. 131. 127.  76.
  58. 361. 409. 224. 175.  37.  69. 206. 191. 182. 410. 185. 176. 212.
 118. 190. 257. 223. 248. 177. 117. 188. 179. 181. 123. 407. 105.  45.
 107. 286.  46. 226. 120. 296. 200.  49.  59.  56.  29.  57.  74.  50.
  35.  43.  78.  54. 129.  99. 435. 104. 319.  27.  97. 293.  79. 219.
 102. 316. 389.  26. 368. 246. 400. 495. 372. 446. 211.  68.  95. 561.
 343. 106. 386. 

In [10]:
# for each column
for col in cols:
    print(col)

    # get a list of unique values
    unique = df[col].unique()

    # if number of unique values is less than 30, print the values. Otherwise print the number of unique values
    if len(unique)<30:
        print(unique, '\n====================================\n\n')
    else:
        print(str(len(unique)) + ' unique values', '\n====================================\n\n')

Menu_Item_ID
44936 unique values 


Year
[2017 2018] 


Restaurant_Item_Name
51660 unique values 


restaurant
96 unique values 


Restaurant_ID
94 unique values 


Item_Name
46995 unique values 


Item_Description
62859 unique values 


Food_Category
['Beverages' 'Entrees' 'Pizza' 'Sandwiches' 'Burgers' 'Salads'
 'Appetizers & Sides' 'Baked Goods' 'Desserts' 'Soup'
 'Toppings & Ingredients' 'Fried Potatoes'] 


Serving_Size
1164 unique values 


Serving_Size_text
[nan "'5-6" '18-25' "'12-20" '16-24' '10.98-11.00' '14.00-56.00'
 '7.00-28.00' '14.00-54.00' '138-139' '12.00-20.00' '620.00-1350.00' 'oz'
 '5.00-6.00' '16.00-24.00' '300.00-680.00'] 


Serving_Size_Unit
[nan 'fl oz' 'g' 'oz' 'ml'] 


Serving_Size_household
816 unique values 


Calories
1348 unique values 


Total_Fat
613 unique values 


Saturated_Fat
391 unique values 


Trans_Fat
98 unique values 


Cholesterol
766 unique values 


Sodium
2428 unique values 


Potassium
196 unique values 


Carbohydrates
718 unique values 

##  The number of missing values of each column.

In [11]:
df.isnull().sum()

Menu_Item_ID                  0
Year                          0
Restaurant_Item_Name          0
restaurant                    0
Restaurant_ID                 0
Item_Name                     0
Item_Description              0
Food_Category                 0
Serving_Size              38320
Serving_Size_text         65180
Serving_Size_Unit         38292
Serving_Size_household    49981
Calories                   9904
Total_Fat                 10373
Saturated_Fat             11076
Trans_Fat                 13716
Cholesterol               12000
Sodium                    10228
Potassium                 64121
Carbohydrates             10931
Protein                   10986
Sugar                     12288
Dietary_Fiber             11779
Calories_100g             39341
Total_Fat_100g            39533
Saturated_Fat_100g        39934
Trans_Fat_100g            41391
Cholesterol_100g          40093
Sodium_100g               39366
Potassium_100g            64596
Carbohydrates_100g        39627
Protein_

In [12]:
df.describe() # for numeric columns

Menu_Item_ID          Year  Restaurant_ID  Serving_Size      Calories  \
count   65219.000000  65219.000000   65219.000000  26899.000000  55315.000000   
mean   103164.326117   2017.461829      53.060151     92.222657    380.020880   
std     60360.317220      0.498545      27.108196    172.826444    420.620976   
min         7.000000   2017.000000       1.000000      0.000000      0.000000   
25%     39361.500000   2017.000000      30.000000     12.000000    130.000000   
50%    122345.000000   2017.000000      56.000000     24.000000    270.000000   
75%    156752.500000   2018.000000      79.000000     98.000000    500.000000   
max    179840.000000   2018.000000      94.000000   3792.000000  13960.000000   

          Total_Fat  Saturated_Fat     Trans_Fat   Cholesterol        Sodium  \
count  54846.000000   54143.000000  51503.000000  53219.000000  54991.000000   
mean      17.237539       6.488704      0.300858     56.726642    686.570538   
std       25.461848      13.354115     10.957444    105.439070   1063.799544   
min        0.000000       0.000000      0.000000      0.000000      0.000000   
25%        0.500000       0.000000      0.000000      0.000000     80.000000   
50%        9.000000       3.500000      0.000000     20.000000    310.000000   
75%       23.000000       9.000000      0.000000     69.000000    930.000000   
max     1078.000000    2046.000000   1470.000000   2680.000000  41290.000000   

       ...    Sodium_100g  Potassium_100g  Carbohydrates_100g  Protein_100g  \
count  ...   25853.000000      623.000000        25592.000000   25531.00000   
mean   ...     292.044482      134.134831           15.033565       4.53719   
std    ...    1448.304081      119.938038           20.082218       6.94306   
min    ...       0.000000        0.000000            0.000000       0.00000   
25%    ...      12.000000       37.000000            6.000000       0.00000   
50%    ...      60.000000      117.000000           12.000000       2.00000   
75%    ...     475.000000      198.000000           19.000000       7.00000   
max    ...  209000.000000      561.000000         1800.000000     425.00000   

         Sugar_100g  Dietary_Fiber_100g     Kids_Meal  Limited_Time_Offer  \
count  25207.000000        25391.000000  65219.000000        65219.000000   
mean       7.602095            0.676145      0.057621            0.032445   
std       15.596987            2.964487      0.233028            0.177179   
min        0.000000            0.000000      0.000000            0.000000   
25%        1.000000            0.000000      0.000000            0.000000   
50%        5.000000            0.000000      0.000000            0.000000   
75%       11.000000            1.000000      0.000000            0.000000   
max     1800.000000          400.000000      1.000000            1.000000   

           Regional     Shareable  
count  65219.000000  65219.000000  
mean       0.031969      0.028642  
std        0.175919      0.166799  
min        0.000000      0.000000  
25%        0.000000      0.000000  
50%        0.000000      0.000000  
75%        0.000000      0.000000  
max        1.000000      1.000000  

[8 rows x 30 columns]

In [13]:
df.describe(include='object')  # for object columns

Restaurant_Item_Name restaurant  \
count                                               65219      65219   
unique                                              51660         96   
top     Jack in the Box Fruitwater Fizzy Cherry Lime, ...  Starbucks   
freq                                                    4       4504   

        Item_Name                 Item_Description Food_Category  \
count       65219                            65219         65219   
unique      46995                            62859            12   
top     Diet Coke  Chocolate Chip Cookie, Desserts     Beverages   
freq           42                               10         23396   

       Serving_Size_text Serving_Size_Unit Serving_Size_household  \
count                 39             26927                  15238   
unique                15                 4                    815   
top          14.00-56.00                 g                1 Slice   
freq                  17             10843                   1704   

       Calories_text Total_Fat_text Saturated_Fat_text Trans_Fat_text  \
count            303             69                 50             10   
unique           238             59                 41              7   
top               <5          68-70               '1-3           '1-3   
freq              17              3                  3              3   

       Cholesterol_text Sodium_text Potassium_text Carbohydrates_text  \
count               358          93              2                634   
unique               28          82              2                 75   
top                  <5       14-55        450-470                 <1   
freq                325           2              1                553   

       Protein_text Sugar_text Dietary_Fiber_text  
count           467        591                811  
unique           58         45                 32  
top              <1         <1                 <1  
freq            406        536                749

In [14]:
df.describe(include='all')  # for both object and numeric data

Menu_Item_ID          Year  \
count    65219.000000  65219.000000   
unique            NaN           NaN   
top               NaN           NaN   
freq              NaN           NaN   
mean    103164.326117   2017.461829   
std      60360.317220      0.498545   
min          7.000000   2017.000000   
25%      39361.500000   2017.000000   
50%     122345.000000   2017.000000   
75%     156752.500000   2018.000000   
max     179840.000000   2018.000000   

                                     Restaurant_Item_Name restaurant  \
count                                               65219      65219   
unique                                              51660         96   
top     Jack in the Box Fruitwater Fizzy Cherry Lime, ...  Starbucks   
freq                                                    4       4504   
mean                                                  NaN        NaN   
std                                                   NaN        NaN   
min                                                   NaN        NaN   
25%                                                   NaN        NaN   
50%                                                   NaN        NaN   
75%                                                   NaN        NaN   
max                                                   NaN        NaN   

        Restaurant_ID  Item_Name                 Item_Description  \
count    65219.000000      65219                            65219   
unique            NaN      46995                            62859   
top               NaN  Diet Coke  Chocolate Chip Cookie, Desserts   
freq              NaN         42                               10   
mean        53.060151        NaN                              NaN   
std         27.108196        NaN                              NaN   
min          1.000000        NaN                              NaN   
25%         30.000000        NaN                              NaN   
50%         56.000000        NaN                              NaN   
75%         79.000000        NaN                              NaN   
max         94.000000        NaN                              NaN   

       Food_Category  Serving_Size Serving_Size_text  ... Sodium_text  \
count          65219  26899.000000                39  ...          93   
unique            12           NaN                15  ...          82   
top        Beverages           NaN       14.00-56.00  ...       14-55   
freq           23396           NaN                17  ...           2   
mean             NaN     92.222657               NaN  ...         NaN   
std              NaN    172.826444               NaN  ...         NaN   
min              NaN      0.000000               NaN  ...         NaN   
25%              NaN     12.000000               NaN  ...         NaN   
50%              NaN     24.000000               NaN  ...         NaN   
75%              NaN     98.000000               NaN  ...         NaN   
max              NaN   3792.000000               NaN  ...         NaN   

       Potassium_text  Carbohydrates_text  Protein_text  Sugar_text  \
count               2                 634           467         591   
unique              2                  75            58          45   
top           450-470                  <1            <1          <1   
freq                1                 553           406         536   
mean              NaN                 NaN           NaN         NaN   
std               NaN                 NaN           NaN         NaN   
min               NaN                 NaN           NaN         NaN   
25%               NaN                 NaN           NaN         NaN   
50%               NaN                 NaN           NaN         NaN   
75%               NaN                 NaN           NaN         NaN   
max               NaN                 NaN           NaN         NaN   

        Dietary_Fiber_text     Kids_Meal  Limited_Time_Offer      Regional  \
count                  811  65219.0000

## Frequency distribution for each column

In [15]:
for col in cols:
    print(col)
    value_counts = df[col].value_counts()
    print(value_counts, '\n\n=======================================================')

Menu_Item_ID
19067     2
45721     2
45760     2
45754     2
45741     2
         ..
156586    1
156584    1
156581    1
140791    1
174902    1
Name: Menu_Item_ID, Length: 44936, dtype: int64 

Year
2017    35099
2018    30120
Name: Year, dtype: int64 

Restaurant_Item_Name
Jack in the Box Fruitwater Fizzy Cherry Lime, Lemonade, Lemon Lime, Mixed Berry, Peach Lemonade or S    4
Jack in the Box POWERade Zero Cherry, Fruit Punch, Grape, Lemon, Lime, Orange, Raspberry or Strawber    4
Jack in the Box Fanta Zero Cherry, Fruit Punch, Grape, Lime, Orange, Peach, Raspberry or Strawberry,    4
Jack in the Box Hi C Cherry, Fruit Punch, Grape, Lemon, Orange, Orange Vanilla, Raspberry, Raspberry    4
BJ's Restaurant & Brewhouse BJs Signature Five Cheese Blend, for 1 Topping Build Your Own Deep Dish     4
                                                                                                       ..
Red Lobster Shrimp & Wood Grilled Chicken w/ Hand Breaded Shrimp                        

Name: Protein_text, dtype: int64 

Sugar_text
<1         536
'5-19        2
0-25         2
'5-18        2
'6-25        2
'7-26        2
'8-30        2
'6-26        2
'6-24        2
'7-27        2
'6-22        2
'9-11        2
'1-6         1
'5-17        1
'2-8         1
0-55         1
'1-2         1
0-1          1
18-21        1
'2-3         1
'6-42        1
19-38        1
'3-28        1
40-61        1
'4-28        1
'8-9         1
0-47         1
34-65        1
'1-30        1
'3-5         1
49-85        1
146-254      1
20-23        1
41-47        1
49-51        1
'5-6         1
'5-14        1
'8-11        1
'4-30        1
'6-45        1
0-30         1
'5-8         1
'7-10        1
0-41         1
'2-35        1
Name: Sugar_text, dtype: int64 

Dietary_Fiber_text
<1        749
'8-9        9
-1          6
0-1         5
'7-8        4
'1-2        3
'5-6        3
'6-8        2
'2-4        2
'4-5        2
<1-1        2
'9-11       2
'10-12      2
'1-4        2
<1-4        1
'2-3        1
'2-

### Checking for duplicate rows

In [16]:
df.duplicated()

0        False
1        False
2        False
3        False
4        False
         ...  
65214    False
65215    False
65216    False
65217    False
65218    False
Length: 65219, dtype: bool

In [17]:
# Selecting duplicate based on all columns
duplicate = df[df.duplicated()]
  
print("Duplicate Rows :")
duplicate

Duplicate Rows :


Empty DataFrame
Columns: [Menu_Item_ID, Year, Restaurant_Item_Name, restaurant, Restaurant_ID, Item_Name, Item_Description, Food_Category, Serving_Size, Serving_Size_text, Serving_Size_Unit, Serving_Size_household, Calories, Total_Fat, Saturated_Fat, Trans_Fat, Cholesterol, Sodium, Potassium, Carbohydrates, Protein, Sugar, Dietary_Fiber, Calories_100g, Total_Fat_100g, Saturated_Fat_100g, Trans_Fat_100g, Cholesterol_100g, Sodium_100g, Potassium_100g, Carbohydrates_100g, Protein_100g, Sugar_100g, Dietary_Fiber_100g, Calories_text, Total_Fat_text, Saturated_Fat_text, Trans_Fat_text, Cholesterol_text, Sodium_text, Potassium_text, Carbohydrates_text, Protein_text, Sugar_text, Dietary_Fiber_text, Kids_Meal, Limited_Time_Offer, Regional, Shareable]
Index: []

[0 rows x 49 columns]

#### There are no duplicate entries in our dataset

---

## Data Profiling Summary 
### List of the issues that the data contains

1. In intial dataset, there were 65219 rows and 49 columns in our dataset.


2. These columns contain integer, float and categorical values.


3. We also observed unique values in each column.


4. The columns Calories_text, Total_Fat_text, Saturated_Fat_text, Trans_Fat_text, Cholesterol_text, Sodium_text,   Potassium_text,Carbohydrates_text,Protein_text,Sugar_text,Dietary_Fiber_text,Potassium,Potassium_100g,Serving_Size_text,Serving_Size,Serving_Size_Unit,Serving_Size_household do not proide relevant information for the model deployment. So these columns needs to be dropped. 


5. Since the major analysis will revolve around the calories column. The values that were na were dropped from the column.


6. There are missing vlaues in 19 columns.Since these are mostly related to nutrients, the na columns are filled with value zero for analysis.



---
---
---






# Data  Cleansing

### Dropping unwanted columns

In [28]:
# Dropping columns
df1=df.drop(['Calories_text', 'Total_Fat_text',
       'Saturated_Fat_text', 'Trans_Fat_text', 'Cholesterol_text',
       'Sodium_text', 'Potassium_text', 'Carbohydrates_text', 'Protein_text',
       'Sugar_text', 'Dietary_Fiber_text','Potassium','Potassium_100g','Serving_Size_text','Serving_Size','Serving_Size_Unit',
       'Serving_Size_household'],axis=1)

## Rename 'df1' columns
By making all the columns’ names in lowercase, and if the column name has a whitespace, replace it with an underscore. (For example: “Customer Type” will be renamed to “customer_type”).

In [29]:
df1.columns

Index(['Menu_Item_ID', 'Year', 'Restaurant_Item_Name', 'restaurant',
       'Restaurant_ID', 'Item_Name', 'Item_Description', 'Food_Category',
       'Calories', 'Total_Fat', 'Saturated_Fat', 'Trans_Fat', 'Cholesterol',
       'Sodium', 'Carbohydrates', 'Protein', 'Sugar', 'Dietary_Fiber',
       'Calories_100g', 'Total_Fat_100g', 'Saturated_Fat_100g',
       'Trans_Fat_100g', 'Cholesterol_100g', 'Sodium_100g',
       'Carbohydrates_100g', 'Protein_100g', 'Sugar_100g',
       'Dietary_Fiber_100g', 'Kids_Meal', 'Limited_Time_Offer', 'Regional',
       'Shareable'],
      dtype='object')

In [30]:
df1.rename(columns = lambda x: x.lower().replace(' ','_'), inplace= True)
df1.columns

Index(['menu_item_id', 'year', 'restaurant_item_name', 'restaurant',
       'restaurant_id', 'item_name', 'item_description', 'food_category',
       'calories', 'total_fat', 'saturated_fat', 'trans_fat', 'cholesterol',
       'sodium', 'carbohydrates', 'protein', 'sugar', 'dietary_fiber',
       'calories_100g', 'total_fat_100g', 'saturated_fat_100g',
       'trans_fat_100g', 'cholesterol_100g', 'sodium_100g',
       'carbohydrates_100g', 'protein_100g', 'sugar_100g',
       'dietary_fiber_100g', 'kids_meal', 'limited_time_offer', 'regional',
       'shareable'],
      dtype='object')

In [31]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65219 entries, 0 to 65218
Data columns (total 32 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   menu_item_id          65219 non-null  int64  
 1   year                  65219 non-null  int64  
 2   restaurant_item_name  65219 non-null  object 
 3   restaurant            65219 non-null  object 
 4   restaurant_id         65219 non-null  int64  
 5   item_name             65219 non-null  object 
 6   item_description      65219 non-null  object 
 7   food_category         65219 non-null  object 
 8   calories              55315 non-null  float64
 9   total_fat             54846 non-null  float64
 10  saturated_fat         54143 non-null  float64
 11  trans_fat             51503 non-null  float64
 12  cholesterol           53219 non-null  float64
 13  sodium                54991 non-null  float64
 14  carbohydrates         54288 non-null  float64
 15  protein            


# Handle missing values
---
### There are a few options for handling missing values:

    - Remove any rows that have missing values.
    - Remove any columns that have missing values.
    - Fill the missing values with some other value.
    - Leave the missing values as is.

The first two options are often used to prepare data for machine learning algorithms, which are unable to be used with data that includes null values. We can use the **DataFrame.dropna()** method to remove or drop rows and columns with null values.

Syntax:

#### DataFrame.dropna(axis=0, how='any',  inplace=False)

Arguments:
###### axis: Default – 0
   -  0, or ‘index’ : Drop rows which contain NaN values.
   - 1, or ‘columns’ : Drop columns which contain NaN value.
    
###### how: Default – ‘any’
   - ‘any’ : Drop rows / columns which contain any NaN values.
   - ‘all’ : Drop rows / columns which contain all NaN values.
    
###### inplace (bool): Default- False
   - If True, modifies the calling dataframe object

In [32]:
# Number of missing values in each coulmn
df1.isnull().sum()

menu_item_id                0
year                        0
restaurant_item_name        0
restaurant                  0
restaurant_id               0
item_name                   0
item_description            0
food_category               0
calories                 9904
total_fat               10373
saturated_fat           11076
trans_fat               13716
cholesterol             12000
sodium                  10228
carbohydrates           10931
protein                 10986
sugar                   12288
dietary_fiber           11779
calories_100g           39341
total_fat_100g          39533
saturated_fat_100g      39934
trans_fat_100g          41391
cholesterol_100g        40093
sodium_100g             39366
carbohydrates_100g      39627
protein_100g            39688
sugar_100g              40012
dietary_fiber_100g      39828
kids_meal                   0
limited_time_offer          0
regional                    0
shareable                   0
dtype: int64

### Dropping the rows which has missing value in calorie column

In [34]:
df1.dropna(subset=['calories'], inplace=True)

In [35]:
df1.isnull().sum()

menu_item_id                0
year                        0
restaurant_item_name        0
restaurant                  0
restaurant_id               0
item_name                   0
item_description            0
food_category               0
calories                    0
total_fat                 550
saturated_fat            1254
trans_fat                3900
cholesterol              2185
sodium                    406
carbohydrates            1088
protein                  1169
sugar                    2443
dietary_fiber            1963
calories_100g           29437
total_fat_100g          29659
saturated_fat_100g      30060
trans_fat_100g          31529
cholesterol_100g        30226
sodium_100g             29502
carbohydrates_100g      29742
protein_100g            29816
sugar_100g              30127
dietary_fiber_100g      29960
kids_meal                   0
limited_time_offer          0
regional                    0
shareable                   0
dtype: int64

#### Fill the missing values with some other value. In this case it has been filled with zero.


In [36]:
df1 = df1.fillna(0)

In [37]:
df1.isnull().sum()

menu_item_id            0
year                    0
restaurant_item_name    0
restaurant              0
restaurant_id           0
item_name               0
item_description        0
food_category           0
calories                0
total_fat               0
saturated_fat           0
trans_fat               0
cholesterol             0
sodium                  0
carbohydrates           0
protein                 0
sugar                   0
dietary_fiber           0
calories_100g           0
total_fat_100g          0
saturated_fat_100g      0
trans_fat_100g          0
cholesterol_100g        0
sodium_100g             0
carbohydrates_100g      0
protein_100g            0
sugar_100g              0
dietary_fiber_100g      0
kids_meal               0
limited_time_offer      0
regional                0
shareable               0
dtype: int64

In [38]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 55315 entries, 2 to 65218
Data columns (total 32 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   menu_item_id          55315 non-null  int64  
 1   year                  55315 non-null  int64  
 2   restaurant_item_name  55315 non-null  object 
 3   restaurant            55315 non-null  object 
 4   restaurant_id         55315 non-null  int64  
 5   item_name             55315 non-null  object 
 6   item_description      55315 non-null  object 
 7   food_category         55315 non-null  object 
 8   calories              55315 non-null  float64
 9   total_fat             55315 non-null  float64
 10  saturated_fat         55315 non-null  float64
 11  trans_fat             55315 non-null  float64
 12  cholesterol           55315 non-null  float64
 13  sodium                55315 non-null  float64
 14  carbohydrates         55315 non-null  float64
 15  protein            

In [39]:
# Reset the index of the DataFrame
df1.reset_index(drop=True, inplace = True)
df1

menu_item_id  year                               restaurant_item_name  \
0             35027  2017                 7 Eleven French Vanilla Cappuccino   
1             35028  2017                          7 Eleven Peppermint Mocha   
2             35029  2017                       7 Eleven Pumpkin Spice Latte   
3             35037  2017                         7 Eleven Caramel Macchiato   
4             35042  2017                             7 Eleven Hot Chocolate   
...             ...   ...                                                ...   
55310        118344  2018               Einstein Bros Frozen Mocha, 24 fl oz   
55311        153014  2018  Dunkin' Donuts Sweet Black Pepper Bacon Wake U...   
55312        174902  2018                  LongHorn Steakhouse Whiskey Smash   
55313        131673  2018                    Golden Corral Chicken Enchilada   
55314        136001  2018                 Del Taco Street Tacos, Carne Asada   

                restaurant  restaurant_id  \
0                 7 Eleven              1   
1                 7 Eleven              1   
2                 7 Eleven              1   
3                 7 Eleven              1   
4                 7 Eleven              1   
...                    ...            ...   
55310        Einstein Bros             31   
55311       Dunkin' Donuts             30   
55312  LongHorn Steakhouse             53   
55313        Golden Corral             37   
55314             Del Taco             27   

                                   item_name  \
0                  French Vanilla Cappuccino   
1                           Peppermint Mocha   
2                        Pumpkin Spice Latte   
3                          Caramel Macchiato   
4                              Hot Chocolate   
...                                      ...   
55310                 Frozen Mocha, 24 fl oz   
55311  Sweet Black Pepper Bacon Wake Up Wrap   
55312                          Whiskey Smash   
55313                      Chicken Enchilada   
55314              Street Tacos, Carne Asada   

                                        item_description food_category  \
0      French Vanilla Cappuccino, Coffee, Drinks, Fla...     Beverages   
1                      Peppermint Mocha, Coffee, 8 fl oz     Beverages   
2                   Pumpkin Spice Latte, Coffee, 8 fl oz     Beverages   
3      Caramel Macchiato, also called Caramel Macchia...     Beverages   
4      Hot Chocolate, Coffee, Drinks, Flavors Vary by...     Beverages   
...                                                  ...           ...   
55310  Frozen Mocha w/ 2% Milk, 24 fl oz, Classic, Be...     Beverages   
55311  Sweet Black Pepper Bacon Wake Up Wrap, Breakfa...    Sandwiches   
55312  Whiskey Smash w/ Jim Beam Bourbon, Fresh Squee...     Beverages   
55313                Chicken Enchilada, Chicken & Turkey       Entrees   
55314  Street Tacos in 2 Tortillas w/ Carne Asada Ste...    Sandwiches   

       calories  total_fat  ...  cholesterol_100g  sodium_100g  \
0         190.0        6.0  ...               0.0         84.0   
1         180.0        4.0  ...               0.0         79.0   
2         190.0        6.0  ...               0.0         84.0   
3         190.0        5.0  ...               0.0         75.0   
4         170.0        2.5  ...               0.0         88.0   
...         ...        ...  ...               ...          ...   
55310     680.0       16.0  ...               7.0         25.0   
55311     250.0       15.0  ...               0.0          0.0   
55312     200.0        0.0  ...               0.0          0.0   
55313     190.0       11.0  ...              31.0        528.0   
55314     180.0        7.0  ...              22.0        354.0   

       carbohydrates_100g  protein_100g  sugar_100g  dietary_fiber_100g  \
0                    15.0           1.0        10.0                 0.0   
1                    16.0           0.0        11.0                 0.0   
2                    

#### Unique values for each column after cleansing

In [40]:
# for each column
cols = df1.columns
for col in cols:
    print(col)

    # get a list of unique values
    unique = df1[col].unique()
    print(sorted(unique), '\n*************************************\n\n')

IOPub data rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_data_rate_limit`.

Current values:
NotebookApp.iopub_data_rate_limit=1000000.0 (bytes/sec)
NotebookApp.rate_limit_window=3.0 (secs)



In [41]:
df1

menu_item_id  year                               restaurant_item_name  \
0             35027  2017                 7 Eleven French Vanilla Cappuccino   
1             35028  2017                          7 Eleven Peppermint Mocha   
2             35029  2017                       7 Eleven Pumpkin Spice Latte   
3             35037  2017                         7 Eleven Caramel Macchiato   
4             35042  2017                             7 Eleven Hot Chocolate   
...             ...   ...                                                ...   
55310        118344  2018               Einstein Bros Frozen Mocha, 24 fl oz   
55311        153014  2018  Dunkin' Donuts Sweet Black Pepper Bacon Wake U...   
55312        174902  2018                  LongHorn Steakhouse Whiskey Smash   
55313        131673  2018                    Golden Corral Chicken Enchilada   
55314        136001  2018                 Del Taco Street Tacos, Carne Asada   

                restaurant  restaurant_id  \
0                 7 Eleven              1   
1                 7 Eleven              1   
2                 7 Eleven              1   
3                 7 Eleven              1   
4                 7 Eleven              1   
...                    ...            ...   
55310        Einstein Bros             31   
55311       Dunkin' Donuts             30   
55312  LongHorn Steakhouse             53   
55313        Golden Corral             37   
55314             Del Taco             27   

                                   item_name  \
0                  French Vanilla Cappuccino   
1                           Peppermint Mocha   
2                        Pumpkin Spice Latte   
3                          Caramel Macchiato   
4                              Hot Chocolate   
...                                      ...   
55310                 Frozen Mocha, 24 fl oz   
55311  Sweet Black Pepper Bacon Wake Up Wrap   
55312                          Whiskey Smash   
55313                      Chicken Enchilada   
55314              Street Tacos, Carne Asada   

                                        item_description food_category  \
0      French Vanilla Cappuccino, Coffee, Drinks, Fla...     Beverages   
1                      Peppermint Mocha, Coffee, 8 fl oz     Beverages   
2                   Pumpkin Spice Latte, Coffee, 8 fl oz     Beverages   
3      Caramel Macchiato, also called Caramel Macchia...     Beverages   
4      Hot Chocolate, Coffee, Drinks, Flavors Vary by...     Beverages   
...                                                  ...           ...   
55310  Frozen Mocha w/ 2% Milk, 24 fl oz, Classic, Be...     Beverages   
55311  Sweet Black Pepper Bacon Wake Up Wrap, Breakfa...    Sandwiches   
55312  Whiskey Smash w/ Jim Beam Bourbon, Fresh Squee...     Beverages   
55313                Chicken Enchilada, Chicken & Turkey       Entrees   
55314  Street Tacos in 2 Tortillas w/ Carne Asada Ste...    Sandwiches   

       calories  total_fat  ...  cholesterol_100g  sodium_100g  \
0         190.0        6.0  ...               0.0         84.0   
1         180.0        4.0  ...               0.0         79.0   
2         190.0        6.0  ...               0.0         84.0   
3         190.0        5.0  ...               0.0         75.0   
4         170.0        2.5  ...               0.0         88.0   
...         ...        ...  ...               ...          ...   
55310     680.0       16.0  ...               7.0         25.0   
55311     250.0       15.0  ...               0.0          0.0   
55312     200.0        0.0  ...               0.0          0.0   
55313     190.0       11.0  ...              31.0        528.0   
55314     180.0        7.0  ...              22.0        354.0   

       carbohydrates_100g  protein_100g  sugar_100g  dietary_fiber_100g  \
0                    15.0           1.0        10.0                 0.0   
1                    16.0           0.0        11.0                 0.0   
2                    

# Export the cleaned dataset to CSV file.

In [43]:
#Exporting data into csv file
df1.to_csv("NYC_Menu.csv", index= False)